### PR TITLE
feat(developer): improve BCP 47 canonicalization

### PR DESCRIPTION
**Pull request by mcdurdin**
Ensures we get a canonical tag per langtags.json as far as we possibly can. This is a breaking change for the compiler as tags which were formerly regarded as canonical are no longer regarded that way. This mostly relates to script subtag but a secondary bug meant that some other tags would have lost data in the canonicalization process (because we did a lookup based only on the language subtag previously, which is a no-no).

See keymanapp/keyboards#1452 for related work.

@ermshiperete this will impact your recent work on BCP47 in the Linux codebase.

### CODE REVIEW

File: windows/src/developer/kmcomp/Keyman.Developer.System.ValidateRepoChanges.pas [hunk 1]
```diff
@@ -0,0 +1,463 @@
+unit Keyman.Developer.System.ValidateRepoChanges;
```
Comment (mcdurdin) [inline windows/src/developer/kmcomp/Keyman.Developer.System.ValidateRepoChanges.pas:1]: This unit is really for internal use only -- I included it in kmcomp at present for testing and we may remove it again later. I don't want to spend effort documenting it at present :)

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas [hunk 2]
```diff
@@ -37,6 +37,10 @@ class function TCanonicalLanguageCodeUtils.FindBestTag(const Tag: string; AddReg
     if t.Tag = '' then
       Exit('');
 
+    // Special case for IPA keyboards; otherwise we'd have und-Zyyy-fonipa
+    if (t.Language = 'und') and (t.Variant = 'fonipa') then
+      Exit('und-fonipa');
+
```
Comment (mcdurdin) [inline windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:43]: @ermshiperete changes to algorithm

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas [hunk 3]
```diff
@@ -47,22 +51,22 @@ class function TCanonicalLanguageCodeUtils.FindBestTag(const Tag: string; AddReg
       t.Tag := Result;
     end;
 
-    if not TLangTagsMap.LangTags.TryGetValue(t.Language, LangTag) then
+    if not TLangTagsMap.LangTags.TryGetValue(t.Tag, LangTag) then
```
Comment (mcdurdin) [inline windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:54]: @ermshiperete changes to algorithm

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas [hunk 4]
```diff
@@ -47,22 +51,22 @@ class function TCanonicalLanguageCodeUtils.FindBestTag(const Tag: string; AddReg
...
+      // We'll make no further assumptions
       Exit(t.Tag);
     end;
 
     // Then, lookup the lang-script and see if there is a suppress-script
-    if SameText(t.Script, LangTag.script) and LangTag.suppress then
-      t.Script := ''
     // Or add the default script in if it is missing and not a suppress-script
-    else if (t.Script = '') and not LangTag.suppress then
+    if (t.Script = '') and not LangTag.suppress and AddScriptIfNotSuppressed then
```
Comment (mcdurdin) [inline windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:63]: @ermshiperete changes to algorithm

File: windows/src/unit-tests/keyboard-info/test-valid.keyboard_info [hunk 5]
```diff
@@ -12,9 +12,9 @@
     },
     "license": "mit",
     "languages": [
-        "bal-Arab",
-        "bgn-Arab",
-        "bgp-Arab"
+        "bal",
+        "bgn",
+        "bgp"
```
Comment (mcdurdin) [inline windows/src/unit-tests/keyboard-info/test-valid.keyboard_info:17]: With the improved canonicalization, script should not be present unless it is needed for disambiguation.

File: windows/src/unit-tests/standards-data/Keyman.Test.System.CanonicalLanguageCodeUtilsTest.pas [hunk 6]
```diff
@@ -23,64 +23,67 @@ implementation
...
+  Assert.AreEqual('sqt-Arab-YE', TCanonicalLanguageCodeUtils.FindBestTag('sqt', True, True));
+  Assert.AreEqual('sqt-Arab-YE', TCanonicalLanguageCodeUtils.FindBestTag('sqt-YE', True, True));
+  Assert.AreEqual('sqt-Arab-YE', TCanonicalLanguageCodeUtils.FindBestTag('sqt-Arab', True, True));
+  Assert.AreEqual('sqt-Latn-YE', TCanonicalLanguageCodeUtils.FindBestTag('sqt-Latn', True, True));
+  Assert.AreEqual('sqt-Latn-YE', TCanonicalLanguageCodeUtils.FindBestTag('sqt-Latn-YE', True, True));
 
-  Assert.AreEqual('sa-Latn-IN', TCanonicalLanguageCodeUtils.FindBestTag('sa-Latn', True));
-  Assert.AreEqual('hi-Latn-IN', TCanonicalLanguageCodeUtils.FindBestTag('hi-Latn', True));
+  Assert.AreEqual('sa-Latn', TCanonicalLanguageCodeUtils.FindBestTag('sa-Latn', True, True));
+  Assert.AreEqual('hi-Latn', TCanonicalLanguageCodeUtils.FindBestTag('hi-Latn', True, True));
```
Comment (mcdurdin) [inline windows/src/unit-tests/standards-data/Keyman.Test.System.CanonicalLanguageCodeUtilsTest.pas:41]: Region is not added to unknown BCP47 tags with the new algorithm. If the tag isn't in langtags.json we don't make assumptions about it.

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas [hunk 7]
```diff
@@ -37,6 +37,10 @@ class function TCanonicalLanguageCodeUtils.FindBestTag(const Tag: string; AddReg
     if t.Tag = '' then
       Exit('');
 
+    // Special case for IPA keyboards; otherwise we'd have und-Zyyy-fonipa
+    if (t.Language = 'und') and (t.Variant = 'fonipa') then
+      Exit('und-fonipa');
```
Comment (darcywong00) [inline windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:42]: What about adding this to the fonipa test in CanonicalLanguageCodeUtilsTest.pas ?

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas [hunk 8]
```diff
@@ -24,7 +24,7 @@ implementation
 ///  This will canonicalize known tags, then apply rules to ensure script subtag
 ///  is present if not suppressed, and add a default region if none given.
```
Comment (darcywong00) [inline windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:25]: Can this get updated for AddScriptIfNotSuppressed?

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas [hunk 9]
```diff
@@ -47,22 +51,22 @@ class function TCanonicalLanguageCodeUtils.FindBestTag(const Tag: string; AddReg
...
     end;
 
     // Then, lookup the lang-script and see if there is a suppress-script
-    if SameText(t.Script, LangTag.script) and LangTag.suppress then
-      t.Script := ''
     // Or add the default script in if it is missing and not a suppress-script
-    else if (t.Script = '') and not LangTag.suppress then
+    if (t.Script = '') and not LangTag.suppress and AddScriptIfNotSuppressed then
+      // AddScriptIfNotSuppressed will generally be True for Windows scenarios;
+      // for other systems and for registry systems it will be False
```
Comment (darcywong00) [inline windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:65]: When would Windows use False? Is it a specific list of tags?

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas (hunk 3 above)
Comment (ermshiperete) [reply windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:54]: Nevermind, I think my build tool that creates the map is wrong. It doesn't process `tags`.

File: windows/src/developer/kmcomp/kmcomp.dproj [hunk 10]
```diff
@@ -267,7 +267,18 @@
...
+  Sentry.Client.Console in &apos;..\..\ext\sentry\Sentry.Client.Console.pas&apos;,
+  sentry in &apos;..\..\ext\sentry\sentry.pas&apos;,
+  Keyman.System.KeymanSentryClient in &apos;..\..\global\delphi\general\Keyman.System.KeymanSentryClient.pas&apos;,
+  KeymanPaths in &apos;..\..\global\delphi\general\KeymanPaths.pas&apos;,
+  Keyman.System.CanonicalLanguageCodeUtils in &apos;..\..\global\delphi\general\Keyman.System.CanonicalLanguageCodeUtils.pas&apos;,
+  Keyman.System.Standards.LangTagsRegistry in &apos;..\..\global\delphi\standards\Keyman.System.Standards.LangTagsRegistry.pas&apos;,
+  Keyman.Developer.System.Project.UrlRenderer in &apos;..\TIKE\project\Keyman.Developer.System.Project.UrlRenderer.pas&apos;,
+  Keyman.Developer.System.ValidateRepoChanges in &apos;Keyman.Developer.System.ValidateRepoChanges.pas&apos;;
+
+{$R icons.RES</Form>
```
Comment (ermshiperete) [inline windows/src/developer/kmcomp/kmcomp.dproj:281]: is that lonely brace correct here?

**Review by ermshiperete — APPROVED** [review]

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas (hunk 9 above)
Comment (mcdurdin) [reply windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:65]: Windows never uses false :)

File: windows/src/developer/kmcomp/kmcomp.dproj (hunk 10 above)
Comment (mcdurdin) [reply windows/src/developer/kmcomp/kmcomp.dproj:281]: Good catch. There is a Delphi compiler bug that sometimes corrupts the .dpr and then that information gets thrown into the .dproj as well. I corrected the .dpr but missed the .dproj.

File: windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas (hunk 9 above)
Comment (mcdurdin) [reply windows/src/global/delphi/general/Keyman.System.CanonicalLanguageCodeUtils.pas:70]: Turns out Linux also uses `True` (https://github.com/keymanapp/keyman/pull/4439/files#diff-9d2b72bb38a2d410351c312d1a57d8a2f410572f21f45a8adad1a2ef1fe549abR265)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.239-beta](https://keyman.com/downloads/releases/beta/14.0.239)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.16-alpha](https://keyman.com/downloads/releases/alpha/15.0.16)

**Pull request merged**